src/ape/api/providers.py

```diff
@@ -23,24 +23,40 @@ class TransactionAPI:
     nonce: Optional[int] = None  # NOTE: `Optional` only to denote using default behavior
     value: int = 0
     gas_limit: Optional[int] = None  # NOTE: `Optional` only to denote using default behavior
-    gas_price: Optional[int] = None  # NOTE: `Optional` only to denote using default behavior
     data: bytes = b""
+    type: str = ""
 
     signature: Optional[TransactionSignature] = None
 
     def __post_init__(self):
         if not self.is_valid:
             raise ProviderError("Transaction is not valid.")
 
+    @property
+    def max_fee(self) -> int:
+        """
+        The total amount in fees willing to be spent on a transaction.
+        Override this property as needed, such as for EIP-1559 differences.
+
+        See :class:`~ape_ethereum.ecosystem.StaticFeeTransaction` and
+        :class`~ape_ethereum.ecosystem.DynamicFeeTransaction` as examples.
+        """
+        return 0
+
     @property
     def total_transfer_value(self) -> int:
         """
         The total amount of WEI that a transaction could use.
         Useful for determining if an account balance can afford
         to submit the transaction.
         """
-        # TODO Support EIP-1559
-        return (self.gas_limit or 0) * (self.gas_price or 0) + self.value
+        return self.value + self.max_fee
+
+    def set_defaults(self, provider: "ProviderAPI"):
+        """
+        Set default values in a transaction with the help of the given provider,
+        such as setting the `gas_limit` from an RPC call to ``eth_gasLimit``.
+        """
```

Comment:
Wondering if the dependency injection might be made a little more explicit

Reply:
This is not really "injecting" the provider here, it is setting defaults on the transaction by making calls to the provider.

I don't really like this approach that much though... I just don't know how else to make it so that different transaction types can request different defaults to be set from the provider based on their type / needs, if that makes sense... Something to keep thinking about! Please help.

Reply:
I moved the `set_defaults` to be on the provider. Now the provider just checks the type of the transaction and sets the defaults accordingly... It is an extra `if` statement but I think the design makes more sense, having the provider take the txn as the arg rather than the other way around, I think this at least partially remedies some of your feedback.

Reply:
^ Scratch that lol... I just deleted the `set_defaults()` method for now and do type checking in the accounts api next to all all the stuff before sending a txn.